Background class, and 5 examples.

DragonMoffon opened the pull request:
The push Implements a background class, and BackgroundTexture, BackgroundGroup, and ParallaxGroup. Plus five examples: static Background, scrolling Background, BackgroundGroup, ParallaxGroup, and Blending between Backgrounds.

These close #1152 and #991

einarf committed:
Shadertoy: Add delta_time property variant

einarf committed:
Update release notes

pvcraven committed:
Work on shadertoy particle tutorial

pvcraven committed:
Merge branch 'development' of github.com:pythonarcade/arcade into development

# Conflicts:
#	doc/tutorials/shader_toy_particles/index.rst

einarf committed:
Bump pyglet to 2.0.dev14

DragonMoffon committed:
Implemented Backgrounds. Now writing all of the examples. background_stationary.py is complete.

DragonMoffon committed:
background_scrolling.py completed.

DragonMoffon committed:
All examples completed.

DragonMoffon committed:
Cleanup for flake 8, and mypy

DragonMoffon committed:
Merge branch 'pythonarcade:development' into backgrounds

einarf commented on arcade/__init__.py:
```diff
@@ -294,6 +294,11 @@ def configure_logging(level: int = None):
 from .sprite_list import get_sprites_at_exact_point
 from .sprite_list import get_sprites_at_point
 
+from .backgrounds import BackgroundTexture
+from .backgrounds import Background
+from .backgrounds import BackgroundGroup
+from .backgrounds import ParallaxGroup
```
I think we should stop adding things to the arcade namespace. It's getting way too cluttered. Users having to import from the `arcade.backgrounds` module is fine.

einarf commented on arcade/backgrounds.py:
```diff
@@ -0,0 +1,408 @@
+from pyglet.math import Mat3
+
+
+class BackgroundTexture:
```
Possibly break this module into smaller parts so we can build on it easier. `arcade/backgrounsd` directly with an `__init__py` importing the classes located in sub-modules.

einarf commented on arcade/backgrounds.py:
```diff
@@ -0,0 +1,408 @@
+    """
+
+    def __init__(self):
+        self.backgrounds: list[Background] = []
```
It's probably better to keep this as a private member and expose a backgrounds property with proper docstrings.

einarf commented on arcade/resources/shaders/background_fs.glsl:
```diff
@@ -0,0 +1,26 @@
+out vec4 fragColor;
+
+void main() {
+    vec2 texSize = vec2(textureSize(backgroundTexture, 0));
```
Implicit conversion from int to float

einarf commented on arcade/resources/shaders/background_fs.glsl:
```diff
@@ -0,0 +1,26 @@
+    vec2 adjusted = (pixelTransform * vec3(adjustedUV, 1.0)).xy;
+
+    adjusted = adjusted / texSize;
+    fragColor = texture(backgroundTexture, adjusted, 0);
```
Bias argument it optional. Just leave it out

einarf commented:
Some quick intial comments. This actually looks really solid, but I am not 100% sure about the details around what properties should be on the texture and the background itself to give the most flexbility.

Also there are missing some type annotations here, but I think fixing the more obvious things and just merging it is a better approach here. Play around with the feature and tweak things instead.

Adding a RGBA color multiplier can also add a lot of flexiblity here instead of just blend. We can really merge this as it is now to be honest. Knitpicking on smaller things that are subject to change around a little bit won't move us forward.

DragonMoffon committed:
Create Background File and __init__.py

DragonMoffon committed:
Create background_texture.py

DragonMoffon committed:
Create background.py

DragonMoffon committed:
Create groups.py

DragonMoffon commented:
For the colour should, do you think I should apply via multiply or addition? Currently, have it as multiply, but if someone wanted to fade in the color, it would be easier if it was added. 

Also, I have it supporting both normalised and not, but I decide if the colour is normalised by seeing if the sum of the colour is less than or equal to 3, which may cause some confusing bugs. Is there a better way of deciding?

einarf commented:
* We do color multiplication everywhere in arcade right now, so that would make the most sense
* If you want to support byte RGBA and normalized RGBA color they can have different properties. `color`, `color_normalized`

DragonMoffon committed:
Added feedback from Einarf. Color in background, a bit more type annotation. made backgrounds in BackgroundGroup and ParallaxGroup 'private'. Gave them __get_item__, __set_item__, and __iter__.

The color was added as RGB with blending separate. Mainly because it makes the purpose of the variable clearer. It is meant to be used for blending between backgrounds, as such I think it should stay as it is. This can be changed if people disagree.

DragonMoffon committed:
Fixed colors.

DragonMoffon commented:
Some of the functions are still missing docstrings. Which will need to be fixed before 2.7 is out. But otherwise, it should be mergeable.
